### PR TITLE
Implement Ratetimeintransit and Shoptimeintransit requests

### PR DESCRIPTION
**Pull request by mazzarito**
I have not taken the time to investigate 100%, but due to what I believe is bad API design on the part of UPS, the entity EstimatedArrival means different things in the rate request context and the time in transit request context. Since these methods combine the two we are left with the options of... 

-  Modifying the current entity so that can handle both types of EstimatedArrival data (this is what I did for now, the downfall is that depending on the requests you have some null fields which could be confusing)
- Creating a new entity (would have to be a different name given the current file structure) or 
- Translate the date and time fields into the current EstimatedArrival entity model (which doesn't fit with current design paradigm)

Here is the current rate example modified for a basic `Shoptimeintransit` request (modifications are at the bottom near the _"CHANGED FROM HERE ON"_ comment):

```php
$rate = new Ups\Rate(
    $accessKey,
    $userId,
    $password
);

try {
    $shipment = new \Ups\Entity\Shipment();

    $shipperAddress = $shipment->getShipper()->getAddress();
    $shipperAddress->setPostalCode('99205');

    $address = new \Ups\Entity\Address();
    $address->setPostalCode('99205');
    $shipFrom = new \Ups\Entity\ShipFrom();
    $shipFrom->setAddress($address);

    $shipment->setShipFrom($shipFrom);

    $shipTo = $shipment->getShipTo();
    $shipTo->setCompanyName('Test Ship To');
    $shipToAddress = $shipTo->getAddress();
    $shipToAddress->setPostalCode('99205');

    $package = new \Ups\Entity\Package();
    $package->getPackagingType()->setCode(\Ups\Entity\PackagingType::PT_PACKAGE);
    $package->getPackageWeight()->setWeight(10);

    // if you need this (depends of the shipper country)
    $weightUnit = new \Ups\Entity\UnitOfMeasurement;
    $weightUnit->setCode(\Ups\Entity\UnitOfMeasurement::UOM_KGS);
    $package->getPackageWeight()->setUnitOfMeasurement($weightUnit);

    $dimensions = new \Ups\Entity\Dimensions();
    $dimensions->setHeight(10);
    $dimensions->setWidth(10);
    $dimensions->setLength(10);

    $unit = new \Ups\Entity\UnitOfMeasurement;
    $unit->setCode(\Ups\Entity\UnitOfMeasurement::UOM_IN);

    $dimensions->setUnitOfMeasurement($unit);
    $package->setDimensions($dimensions);

    $shipment->addPackage($package);

    /**
     * CHANGED FROM HERE ON 
     */
    $deliveryTimeInformation = new \Ups\Entity\DeliveryTimeInformation();
    $deliveryTimeInformation->setPackageBillType(\Ups\Entity\DeliveryTimeInformation::PBT_NON_DOCUMENT);
    $shipment->setDeliveryTimeInformation($deliveryTimeInformation);

    var_dump($rate->shopRatesTimeInTransit($shipment));
} catch (Exception $e) {
    var_dump($e);
}

### CODE REVIEW

**Comment by stefandoorn** [comment]
I think it would be best to add another Entity, like I did for the AV entities that also got duplicated. When using them (in the implementation class) you can check whether they are using the right one. Anyone will figure out while developing that they need the other one. I think it keeps things more clean and the purpose of entities is more obvious. 

It's a bit of duplication, but you can use some inheritance I think. Besides, it would be something to be fixed on a re-write, as we can't afford to introduce BC I think.

**Comment by mazzarito** [comment]
Alright, I created - `Ups\Entity\Rate\EstimatedArrival`, since it is technically part of the rate request, in keeping with what I believe is your previous structure. Since the implementation class `Ups\Entity\ServiceSummary` instantiates EstimatedArrival, how do you think we handle that?

**Comment by mazzarito** [comment]
@stefandoorn I went ahead and added the new class but I'm not sure how you want to handle updating the `Ups\Entity\ServiceSummary` class or if I should create a new `Ups\Entity\Rate\ServiceSummary` entity.

**Comment by stefandoorn** [comment]
Isn't it possible to extend the 'Ups\Entity\EstimatedArrival' from 'Ups\Entity\Rate\EstimatedArrival' and only implement the changes required on top of it? Besides I think it would be good to add the getters/setters also and make properties not public, to keep things more sane.

I think you have to go the same route for ServiceSummary, only overriding the methods that are different and implementing them in a subclass.

I have more time next week to look better into it, but I think this would do for now.

**Comment by mazzarito** [comment]
Well, we can certainly extend it and I definitely agree with not having public properties (that was how the original class is done, I updated both now) but we run into the same issue of having unused properties either way we go, since at the root of it you're describing 2 different entities (but with the same name, and some intersecting similar properties). 

Check out what I just did and see if you think its better... My personal opinion since all we're doing is describing XML entities I'm not sure there's a ton of mileage to get out of inheritance but I'll leave it to you to make the call.

**Comment by stefandoorn** [comment]
I see. I think to make things less complex is to keep as much duplicate code out as possible, therefore I would go with inheritance. Because the names are already quite conflicting :) I can see from performance etc. we don't get gain a lot. 

The implementing method (inside Rate / TimeInTransit) should check on whether it got the right entity supplied before rendering the XML. So I think this is the way to go for the current version of this library.

Next to this I'm in doubt whether we should stick all inside Rate or introduce a RateTimeTransit class to keep things more separated. Because basically this method is going to do more than just rates, and it could also fit inside TimeInTransit. For me it would make sense to separate it into it's own class.

**Comment by mazzarito** [comment]
Hmm yes it the reason I stuck it in Rate was because it calls the same API endpoint- granted at the time I didn't realize the descrepencies we'd have with the different entity types, etc. Do you think that RateTimeInTransit should extend Rate then?

**Comment by stefandoorn** [comment]
I let that up to your judgement if extending is useful in this case, it's not easy for me to oversee how useful that is. If you can re-use things and because it's the same endpoint, I think I would choose extending. If most methods cannot be re-used, maybe separating things makes more sense.

**Comment by mazzarito** [comment]
@stefandoorn Alright I separated it into another class for now, I did have to change a method and a property to protected though

File: src/Entity/Arrival.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    public function getTime()
+    {
+        return $this->time;
+    }
+
+    /**
+     * @param string $time
+     */
+    public function setTime($time)
```
Comment (stefandoorn) [inline src/Entity/Arrival.php]: Same goes here. Or we could even move them together in a single method/property if they always come together.

File: src/Entity/DeliveryTimeInformation.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        if (null === $document) {
+            $document = new DOMDocument();
+        }
+
+        $node = $document->createElement('DeliveryTimeInformation');
+        $node->appendChild($document->createElement('PackageBillType', $this->getPackageBillType()));
+
+        if ($this->getPickup() !== null) {
+            $node->appendChild($document->createElement('Pickup', $this->getPickup()->toNode()));
```
Comment (stefandoorn) [inline src/Entity/DeliveryTimeInformation.php]: Could you delegate the creation of 'Pickup' to the Pickup object? Looking at NodeInterface (and e.g. Service.php as example) you can send $document as parameter to toNode and do all the generation in there. It offloads all generation to the responsible object.

So something like this:

````
$node->appendChild($this->getPickup()->toNode());
````

And inside Pickup:

````
$node = $document->createElement('Pickup');
// In here adding of elements
````

File: src/RateTimeInTransit.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * @param $rateRequest
+     *
+     * @throws Exception
+     *
+     * @return RateResponse
+     */
+    public function getRateTimeInTransit($rateRequest)
+    {
+        if ($rateRequest instanceof Shipment) {
```
Comment (stefandoorn) [inline src/RateTimeInTransit.php]: Can we skip this part and just require `getRateTimeInTransit` to only accept 'RateRequest' object?

File: src/RateTimeInTransit.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * @param $rateRequest
+     *
+     * @throws Exception
+     *
+     * @return RateResponse
+     */
+    public function shopRatesTimeInTransit($rateRequest)
```
Comment (stefandoorn) [inline src/RateTimeInTransit.php]: Same here

**Comment by stefandoorn** [comment]
Alright, getting close :) I put some small review notes. Besides that I think it would be great to have some tests for this class including requests and responses. Also we needs some documentation to put in the README so users know how to use it. Afterwards I'm happy to merge the PR.

**Comment by ryross** [comment]
Is this pretty close to being done? I'd love to use this in a project if it is! @mazzarito Do you need help finishing this up? I can contribute some time here

File: src/RateTimeInTransit.php (hunk 3 above)
Comment (ryross) [reply src/RateTimeInTransit.php]: I like the idea of specifying the type as a parameter. I think @mazzarito was just copying what's happening in Rate though. Should we keep RateTimeInTransit similar?

Comment (stefandoorn) [reply src/RateTimeInTransit.php]: I see, I can live with that.

**Comment by ryross** [comment]
@mazzarito Do you have time to review my changes to your fork?

**Comment by ryross** [comment]
Ping @mazzarito :)

**Comment by mazzarito** [comment]
Hi guys, sorry for being MIA, will review today =)

**Comment by ryross** [comment]
Thanks @mazzarito! @stefandoorn, have any idea why StyleCI is erring here? Are we good otherwise?

**Comment by stefandoorn** [comment]
Squashed commits and will continue in #193. StyleCI had a temporary error it seems.

**Comment by stefandoorn** [comment]
Merged. Thanks @mazzarito & @ryross! Will file a new release soon.

**Comment by ryross** [comment]
Awesome. Thanks!

**Pull request closed without merging**